File: 19 - Remove Nth Node From End of List.cpp
/*
 19. Remove Nth Node From End of List (Medium)

 Notes:
    fast-slow pointer - O(n)
    Use two pointers to scan the list, when the first pointer points to the Nth node, the second pointer move
    backward from the head and follow the first pointer in the following iteration. When the first pointer
    points to the nullptr, the second pointer is pointing to the n + 1 to last node
*/

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        for (int i = 0; i < n; i++)
            head = head->next;

        ListNode* nth_prev = dummy;
        while (head)
        {
            nth_prev = nth_prev->next;
            head = head->next;
        }

        ListNode* temp = nth_prev->next;
        nth_prev->next = temp->next;
        delete temp;

        return dummy->next;
    }
};

File: JAVA/19 - Remove Nth Node From End of List.java
/*
 19. Remove Nth Node From End of List (M)

 Approach 1 :
 Move ptr1 from head to (n + 1) position, then make ptr2 point to dummy and 
 follow the ptr1, while ptr1 points to null, the ptr2 will point to the previous 
 node of Nthnode from end of list. Remove the node
 Time: O(n)
 Space: O(1)

*/

/* Approach 1 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        for (int i = 0; i < n; i++) 
            head = head.next;        
        
        ListNode prev = dummy;
        
        while (head != null) {
            head = head.next;
            prev = prev.next;
        }
        
        prev.next = prev.next.next;
        return dummy.next; 
    }
}

